cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,1111 @@
+		if ((n < 0) && (errno == EINTR)) {
+			continue;
+		} else if (n < 0) {
+			goto exit;
+		}
+
+		for (i = 0; i < n; i++) {
+			/*
+			 * An error has occured on this fd, or
+			 * the socket is not ready for reading
+			 * (why were we notified then?)
+			 */
+			if (!(events[i].events & EPOLLIN)) {
+				ZREPL_ERRLOG("epoll err() :%d\n", errno);
+				close(events[i].data.fd);
+				/*
+				 * TODO:We have choosen to exit
+				 * instead of continuing here.
+				 */
+				goto exit;
+			}
+			/*
+			 * We have a notification on the listening
+			 * socket, which means one or more incoming
+			 * connections.
+			 */
+#if 0
+			while (1) {
+#endif
+				in_len = sizeof (in_addr);
+				new_fd = accept(events[i].data.fd,
+				    &in_addr, &in_len);
+#if 0
+				if ((errno == EAGAIN) ||
+				    (errno == EWOULDBLOCK)) {
+					break;
+				}
+#endif
+				if (new_fd == -1) {
+					ZREPL_ERRLOG("accept err() :%d\n",
+					    errno);
+					goto exit;
+				}
+#ifdef DEBUG
+				hbuf = kmem_alloc(
+				    sizeof (NI_MAXHOST), KM_SLEEP);
+				sbuf = kmem_alloc(
+				    sizeof (NI_MAXSERV), KM_SLEEP);
+				rc = getnameinfo(&in_addr, in_len, hbuf,
+				    sizeof (hbuf), sbuf, sizeof (sbuf),
+				    NI_NUMERICHOST | NI_NUMERICSERV);
+				if (rc == 0) {
+					ZREPL_LOG("Accepted connection on "
+					    "descriptor %d "
+					    "(host=%s, port=%s)\n",
+					    new_fd, hbuf, sbuf);
+					printf("Accepted IO conn on "
+					    "descriptor %d "
+					    "(host=%s, port=%s)\n",
+					    new_fd, hbuf, sbuf);
+				}
+
+				free(hbuf);
+				free(sbuf);
+#endif
+				int *thread_fd = kmem_alloc(
+				    sizeof (int), KM_SLEEP);
+				*thread_fd = new_fd;
+				thrd_info = zk_thread_create(NULL, 0,
+				    (thread_func_t)uzfs_zvol_io_receiver,
+				    (void *)thread_fd, 0, NULL, TS_RUN, 0,
+				    PTHREAD_CREATE_DETACHED);
+				VERIFY3P(thrd_info, !=, NULL);
+#if 0
+			}
+#endif
+		}
+	}
+exit:
+	if (events != NULL) {
+		free(events);
+	}
+
+	if (sfd != -1) {
+		close(sfd);
+	}
+
+	if (efd != -1) {
+		close(efd);
+	}
+
+	printf("uzfs_zvol_io_conn_acceptor thread exiting\n");
+	ZREPL_ERRLOG("uzfs_zvol_io_conn_acceptor thread exiting\n");
+	zk_thread_exit();
+}
+
+/*
+ * One thread per LUN/vol. This thread works
+ * on queue and it sends ack back to client on
+ * a given fd.
+ */
+static void
+uzfs_zvol_io_ack_sender(void *arg)
+{
+	int fd;
+	zvol_info_t		*zinfo;
+	thread_args_t 		*thrd_arg;
+	zvol_io_cmd_t 		*zio_cmd = NULL;
+
+	thrd_arg = (thread_args_t *)arg;
+	fd = thrd_arg->fd;
+	zinfo = uzfs_zinfo_lookup(thrd_arg->zvol_name);
+	free(arg);
+	while (1) {
+		int rc = 0;
+		(void) pthread_mutex_lock(&zinfo->complete_queue_mutex);
+		do {
+			if (STAILQ_EMPTY(&zinfo->complete_queue)) {
+				zinfo->io_ack_waiting = 1;
+				pthread_cond_wait(&zinfo->io_ack_cond,
+				    &zinfo->complete_queue_mutex);
+
+				zinfo->io_ack_waiting = 0;
+				if ((zinfo->state == ZVOL_INFO_STATE_OFFLINE) ||
+				    (zinfo->conn_closed == true)) {
+					(void) pthread_mutex_unlock(
```

Comment:
this unlock can be moved to the end of 'exit'

Reply:
As of today we have 2 exit in this function and in one the exit path we have lock held and in other exit path we do not have lock. Considering that we can not make it generic to release lock at exit. W.r.t to correctness/race, this thread would not be here unless Producer set conn_closed to true so we are okay to free zio_cmd without lock protection but if you insist I can take lock again and free zio_cmd(Completion Queue) here.